extensions/gitpod/src/authentication.ts

```diff
@@ -0,0 +1,253 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Gitpod. All rights reserved.
+ *--------------------------------------------------------------------------------------------*/
+
+import * as vscode from 'vscode';
+import { v4 as uuid } from 'uuid';
+import { Keychain } from './common/keychain';
+import { GitpodServer } from './gitpodServer';
+import Log from './common/logger';
+import { arrayEquals } from './common/utils';
+import { Disposable } from './common/dispose';
+import TelemetryReporter from './telemetryReporter';
+
+interface SessionData {
+	id: string;
+	account?: {
+		label?: string;
+		displayName?: string;
+		id: string;
+	};
+	scopes: string[];
+	accessToken: string;
+}
+
+export default class GitpodAuthenticationProvider extends Disposable implements vscode.AuthenticationProvider {
+	private _sessionChangeEmitter = new vscode.EventEmitter<vscode.AuthenticationProviderAuthenticationSessionsChangeEvent>();
+	private _logger: Log;
+	private _telemetry: TelemetryReporter;
+	private _gitpodServer: GitpodServer;
+	private _keychain: Keychain;
+
+	private _sessionsPromise: Promise<vscode.AuthenticationSession[]>;
+
+	constructor(private readonly context: vscode.ExtensionContext, logger: Log, telemetry: TelemetryReporter) {
+		super();
+
+		this._logger = logger;
+		this._telemetry = telemetry;
+		this._gitpodServer = new GitpodServer(this._logger);
+		this._keychain = new Keychain(this.context, `gitpod.auth`, this._logger);
```

Comment:
it should be scoped per gitpod host, if `gitpod.host` setting is changed then we should reread sessions for given host and unload for old host.

Reply:
I have now changed this so the key for the keychain is `gitpod.auth.host`.